Sources/TuistSupport/Extensions/ConcurrencyExtensions.swift

```diff
@@ -0,0 +1,21 @@
+import Foundation
+
+/// Runs an `action` that is cancelled after the given `timeout`. When the `timeout` elapses, `onTimeout` callback is called.
+/// Inspired by: https://alejandromp.com/development/blog/the-importance-of-cooperative-cancellation/
+public func withTimeout(
+    _ timeout: Duration,
+    onTimeout: @escaping () -> Void,
+    action: @escaping () async throws -> Void
+) async throws {
+    try await withThrowingTaskGroup(of: Void.self) { group in
+        group.addTask {
+            try await action()
+        }
+        group.addTask {
+            try await Task.sleep(for: timeout)
+            onTimeout()
```

Comment:
I think I'm missing something here. Who is cancelling the tasks when the sleep runs over?

Reply:
If the timeout task is finished first, the action task is cancelled [here](https://github.com/tuist/tuist/pull/7105/files/1a769f93a2845f4a691ff35d49083b1bc1c2e068#diff-e19f94f7e25528f2746b59c7be156a97634f5512e98cbf3a5701d1ad2edb667cR19) (or vice versa).

I recommend reading the blog post I attached [here](https://github.com/tuist/tuist/pull/7105/files/1a769f93a2845f4a691ff35d49083b1bc1c2e068#diff-e19f94f7e25528f2746b59c7be156a97634f5512e98cbf3a5701d1ad2edb667cR4): https://alejandromp.com/development/blog/the-importance-of-cooperative-cancellation/